src/result.h

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) 2022 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_RESULT_H
+#define BITCOIN_RESULT_H
+
+#include <optional>
+#include <util/translation.h>
+#include <variant>
+
+/*
+ * 'BResult' is a generic class useful for wrapping a return object
+ * (in case of success) or propagating the error cause.
+*/
+template<class T>
+class BResult {
+private:
+    std::variant<bilingual_str, T> m_variant;
+
+public:
+    BResult() : m_variant(Untranslated("")) {}
+    BResult(const T& _obj) : m_variant(_obj) {}
+    BResult(const bilingual_str& error) : m_variant(error) {}
+
+    /* In case of success, the result object */
+    const T* GetObj() const { return std::get_if<T>(&m_variant); }
```

Comment:
`std::get` on a variant throws [std::bad_variant_access](https://en.cppreference.com/w/cpp/utility/variant/bad_variant_access) on errors (if the variant isn't storing the type) --> https://en.cppreference.com/w/cpp/utility/variant/get. `std::get_if` just return a nullptr.

Ideally, I think that an assertion here would be good, so we can directly return the const reference instead of the pointer (which will force the callers to always check if the result succeeded or not before getting the type). But.. I didn't add it here to not make this PR too broad, it was needing further restructuring in some of the function's callers (add a new plug-and-play result class is one thing, restructure how each of the callers treats the result is another one).

Reply:
I meant renaming the function name from `GetObj` to `Get`

Reply:
Ah ok, I tried to distinguish clearly at the interface level the two possible outcomes `GetError()`/`GetObj()` (It could had been a `GetRes()` as well). Mainly to prevent this class from be blindly used as a smart pointer, forgetting the error existence.
But if you have a strong predilection for the `Get` wording, could change it. Nothing is set on stone hehe.